openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,435 @@
+
+	// strip any resource annotations specified by a user
+	stripResourcesAnnotations(pod.Annotations)
+	// strip any workload annotation to prevent from underlying components(CRI-O, kubelet) to apply any changes
+	// according to the workload annotation
+	stripWorkloadAnnotations(pod.Annotations)
+
+	// more than one workload annotation exists under the pod and we do not support different workload types
+	// under the same pod
+	if len(workloadAnnotations) > 1 {
+		pod.Annotations[workloadAdmissionWarning] = fmt.Sprintf(admissionWarningMultipleWorkloadAnnotations, workloadAnnotations)
+		return nil
+	}
+
+	workloadType, err := getWorkloadType(workloadAnnotations)
+	if err != nil {
+		pod.Annotations[workloadAdmissionWarning] = fmt.Sprintf(admissionWarningFailedGetWorkloadType, err)
+		return nil
+	}
+
+	// no workload annotation is available
+	if workloadType == "" {
+		return nil
+	}
+
+	workloadAnnotation := fmt.Sprintf("%s%s", podWorkloadAnnotationPrefix, workloadType)
+	// TODO: currently we do not check the effect, but it can be useful in the future
+	_, err = getWorkloadAnnotationEffect(workloadAnnotations[workloadAnnotation])
+	if err != nil {
+		pod.Annotations[workloadAdmissionWarning] = fmt.Sprintf(admissionWarningFailedGetWorkloadAnnotationEffect, err)
+		return nil
+	}
+
+	if !a.waitForSyncedStore(time.After(timeToWaitForCacheSync)) {
+		return admission.NewForbidden(attr, fmt.Errorf("%s node and namespace caches not synchronized", PluginName))
+	}
+
+	// allow annotations on project to override management pods CPUs requests
+	ns, err := a.nsLister.Get(attr.GetNamespace())
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return err
+		}
+
+		return admission.NewForbidden(attr, err) // can happen due to informer latency
+	}
+
+	if !isNamespaceAllowed(ns.Annotations, workloadType) {
+		return nil
+	}
+
+	nodes, err := a.nodeLister.List(labels.Everything())
+	if err != nil {
+		return admission.NewForbidden(attr, err) // can happen due to informer latency
+	}
+
+	if len(nodes) == 0 {
+		return nil
+	}
+
+	// TODO: currently we supports only SNO use case because we have not yet worked out the best approach to determining whether the feature
+	// should be on or off in a multi-node cluster, and computing that state incorrectly could lead to breaking running clusters.
+	if len(nodes) > 1 {
+		return nil
+	}
+
+	if err := isManagementResourceAvailableForAllNodes(nodes, workloadType); err != nil {
+		pod.Annotations[workloadAdmissionWarning] = err.Error()
+		return nil
+	}
+
+	allContainers := append([]coreapi.Container{}, pod.Spec.InitContainers...)
+	allContainers = append(allContainers, pod.Spec.Containers...)
+	podQoSClass := getPodQoSClass(allContainers)
+
+	// we do not want to change guaranteed pods resource allocation, because it should be managed by
+	// relevant managers(CPU and memory) under the kubelet
+	if podQoSClass == coreapi.PodQOSGuaranteed {
+		pod.Annotations[workloadAdmissionWarning] = admissionWarningGuaranteedPod
```

Comment:
which pods fall under this class in a real cluster?

Reply:
From a running cluster:
```
$ kubectl get pods -A -o json | jq '.items[] | select(.status.qosClass == "Guaranteed") | .metadata.namespace + "/" + .metadata.name'                                           sttts-exclude-tech-preview
"openshift-etcd/installer-2-ip-10-0-142-246.ec2.internal"
"openshift-etcd/installer-2-ip-10-0-149-75.ec2.internal"
"openshift-etcd/installer-2-ip-10-0-163-51.ec2.internal"
"openshift-etcd/installer-3-ip-10-0-142-246.ec2.internal"
"openshift-etcd/installer-3-ip-10-0-149-75.ec2.internal"
"openshift-etcd/installer-3-ip-10-0-163-51.ec2.internal"
"openshift-etcd/revision-pruner-2-ip-10-0-142-246.ec2.internal"
"openshift-etcd/revision-pruner-2-ip-10-0-149-75.ec2.internal"
"openshift-etcd/revision-pruner-2-ip-10-0-163-51.ec2.internal"
"openshift-etcd/revision-pruner-3-ip-10-0-142-246.ec2.internal"
"openshift-etcd/revision-pruner-3-ip-10-0-149-75.ec2.internal"
"openshift-etcd/revision-pruner-3-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/installer-10-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/installer-10-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/installer-10-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/installer-11-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/installer-11-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/installer-11-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/installer-12-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/installer-13-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/installer-13-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/installer-13-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/installer-14-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/installer-15-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/installer-15-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/installer-15-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/installer-16-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/installer-16-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/installer-16-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/installer-17-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/installer-17-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/installer-17-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/installer-6-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/installer-6-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/installer-7-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/revision-pruner-10-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/revision-pruner-10-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/revision-pruner-10-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/revision-pruner-11-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/revision-pruner-11-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/revision-pruner-11-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/revision-pruner-13-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/revision-pruner-13-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/revision-pruner-13-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/revision-pruner-15-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/revision-pruner-15-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/revision-pruner-16-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/revision-pruner-16-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/revision-pruner-16-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/revision-pruner-17-ip-10-0-142-246.ec2.internal"
"openshift-kube-apiserver/revision-pruner-17-ip-10-0-149-75.ec2.internal"
"openshift-kube-apiserver/revision-pruner-17-ip-10-0-163-51.ec2.internal"
"openshift-kube-apiserver/revision-pruner-6-ip-10-0-149-75.ec2.internal"
"openshift-kube-controller-manager/installer-3-ip-10-0-142-246.ec2.internal"
"openshift-kube-controller-manager/installer-4-ip-10-0-142-246.ec2.internal"
"openshift-kube-controller-manager/installer-5-ip-10-0-163-51.ec2.internal"
"openshift-kube-controller-manager/installer-6-ip-10-0-142-246.ec2.internal"
"openshift-kube-controller-manager/installer-6-ip-10-0-149-75.ec2.internal"
"openshift-kube-controller-manager/installer-6-ip-10-0-163-51.ec2.internal"
"openshift-kube-controller-manager/installer-7-ip-10-0-142-246.ec2.internal"
"openshift-kube-controller-manager/installer-7-ip-10-0-149-75.ec2.internal"
"openshift-kube-controller-manager/installer-7-ip-10-0-163-51.ec2.internal"
"openshift-kube-controller-manager/installer-8-ip-10-0-142-246.ec2.internal"
"openshift-kube-controller-manager/installer-8-ip-10-0-163-51.ec2.internal"
"openshift-kube-controller-manager/installer-9-ip-10-0-142-246.ec2.internal"
"openshift-kube-controller-manager/installer-9-ip-10-0-149-75.ec2.internal"
"openshift-kube-controller-manager/installer-9-ip-10-0-163-51.ec2.internal"
"openshift-kube-controller-manager/revision-pruner-6-ip-10-0-142-246.ec2.internal"
"openshift-kube-controller-manager/revision-pruner-6-ip-10-0-149-75.ec2.internal"
"openshift-kube-controller-manager/revision-pruner-7-ip-10-0-142-246.ec2.internal"
"openshift-kube-controller-manager/revision-pruner-7-ip-10-0-149-75.ec2.internal"
"openshift-kube-controller-manager/revision-pruner-7-ip-10-0-163-51.ec2.internal"
"openshift-kube-controller-manager/revision-pruner-8-ip-10-0-142-246.ec2.internal"
"openshift-kube-controller-manager/revision-pruner-9-ip-10-0-142-246.ec2.internal"
"openshift-kube-controller-manager/revision-pruner-9-ip-10-0-149-75.ec2.internal"
"openshift-kube-controller-manager/revision-pruner-9-ip-10-0-163-51.ec2.internal"
"openshift-kube-scheduler/installer-5-ip-10-0-142-246.ec2.internal"
"openshift-kube-scheduler/installer-6-ip-10-0-142-246.ec2.internal"
"openshift-kube-scheduler/installer-7-ip-10-0-149-75.ec2.internal"
"openshift-kube-scheduler/installer-7-ip-10-0-163-51.ec2.internal"
"openshift-kube-scheduler/installer-8-ip-10-0-142-246.ec2.internal"
"openshift-kube-scheduler/installer-8-ip-10-0-149-75.ec2.internal"
"openshift-kube-scheduler/installer-8-ip-10-0-163-51.ec2.internal"
"openshift-kube-scheduler/installer-9-ip-10-0-142-246.ec2.internal"
"openshift-kube-scheduler/installer-9-ip-10-0-149-75.ec2.internal"
"openshift-kube-scheduler/installer-9-ip-10-0-163-51.ec2.internal"
"openshift-kube-scheduler/revision-pruner-7-ip-10-0-163-51.ec2.internal"
"openshift-kube-scheduler/revision-pruner-8-ip-10-0-142-246.ec2.internal"
"openshift-kube-scheduler/revision-pruner-8-ip-10-0-149-75.ec2.internal"
"openshift-kube-scheduler/revision-pruner-9-ip-10-0-142-246.ec2.internal"
"openshift-kube-scheduler/revision-pruner-9-ip-10-0-149-75.ec2.internal"
"openshift-kube-scheduler/revision-pruner-9-ip-10-0-163-51.ec2.internal"
```

These are the manually scheduled pods from static pod operators. Are these expected?